urld2fb1d2b05_thnaeff_Util_tgz-pJ8-ch_thn_util_tests_TreeTestJ8/src/ch/thn/util/file/PollingWatchService.java

```diff
@@ -0,0 +1,456 @@
+import java.nio.file.Watchable;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import ch.thn.util.thread.ControlledRunnable;
+
+/**
+ * An implementation of a {@link WatchService} which periodically checks the registered
+ * paths for changes.
+ * 
+ * 
+ * 
+ *
+ * @author Thomas Naeff (github.com/thnaeff)
+ *
+ */
+public class PollingWatchService extends ControlledRunnable implements WatchService {
+
+	private long pollTimeout = 0;
+
+	private ConcurrentHashMap<Path, PollingWatchKey> registeredPaths = null;
+
+	/**
+	 * A directory and all the files and their last modified timestamps<br />
+	 * &lt;Directory, &lt;File path, last modified timestamp&gt;&gt;
+	 */
+	private ConcurrentHashMap<Path, Map<Path, Long>> lastModified = null;
+
+	/**
+	 * A FIFO list of all the watch keys which have events pending
+	 */
+	private LinkedBlockingQueue<PollingWatchKey> keysWithEvents = null;
+
+	private FilenameFilter fileNameFilter = null;
+
+	/**
+	 * 
+	 * 
+	 * @param pollTimeout The timeout to wait between two file checks.
+	 */
+	protected PollingWatchService(long pollTimeout) {
+		this(pollTimeout, null);
+	}
+
+	/**
+	 * 
+	 * 
+	 * @param pollTimeout The timeout to wait between two file checks.
+	 * @param fileNameFilter A filter to only check the filtered files. This can
+	 * be a big performance improvements when dealing with large and/or many directories
+	 */
+	protected PollingWatchService(long pollTimeout, FilenameFilter fileNameFilter) {
+		super(true, true);
+
+		this.pollTimeout = pollTimeout;
+		this.fileNameFilter = fileNameFilter;
+
+		registeredPaths = new ConcurrentHashMap<>();
+		lastModified = new ConcurrentHashMap<>();
+		keysWithEvents = new LinkedBlockingQueue<PollingWatchKey>();
+
+	}
+
+
+
+	@Override
+	public void run() {
+		runStart();
+
+		while (!isStopRequested()) {
+			runReset();
+			runPause(false);
+
+			if (isStopRequested()) {
+				break;
+			}
+
+			if (isPauseRequested() || isResetRequested()) {
+				continue;
+			}
+
+			lookForChanges();
+
+			controlledWait(pollTimeout);
+
+		}
+
+		registeredPaths.clear();
+		lastModified.clear();
+		keysWithEvents.clear();
+
+		for (PollingWatchKey key : registeredPaths.values()) {
+			key.cancel();
+		}
+
+		notifyWaitingTake();
+
+		runEnd();
+	}
+
+
+	/**
+	 * 
+	 * 
+	 */
+	private void lookForChanges() {
+
+		//Set<Path> tempPaths = new HashSet<>();
+
+		for (Path directoryPath : registeredPaths.keySet()) {
+			//If a registered path gets deleted because it does not exist any more
+			//and it is the next path in line, it still gets returned. Check here again
+			//if it it still available.
+			if (!registeredPaths.containsKey(directoryPath)) {
+				continue;
+			}
+
+			File directory = directoryPath.toFile();
+			PollingWatchKey key = registeredPaths.get(directoryPath);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **146 and 151**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The removes that can remove the key are at lines: **120** and **233**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).